neuros/main.py

```diff
@@ -58,34 +68,51 @@ def board_stream(board_id: int = BoardIds.SYNTHETIC_BOARD,
 
 
 def stream_windows(board: BoardShim, config: WindowConfig) -> Iterator[np.ndarray]:
+    """
+    Generate windows of EEG data from board stream.
+
+    This generator yields windows of data continuously from the board stream.
+    Each window contains data from all EEG channels and can overlap with previous windows.
+
+    Args:
+        board: Connected and streaming BoardShim instance
+        config: Window configuration for size and overlap
+
+    Yields:
+        Windows of data as numpy arrays with shape (channels, samples)
+    """
+    # Convert ms-based config to samples
     sample_rate = board.get_sampling_rate(board_id=board.board_id)
     window_samples, overlap_samples = config.to_samples(sample_rate)
-    new_samples = window_samples - overlap_samples
+    needed = window_samples - overlap_samples
 
-    # Get the EEG channel indices
+    # Get EEG channel information
     eeg_channels = board.get_eeg_channels(board_id=board.board_id)
 
-    # Create buffer just for EEG channels
-    ring_buffer = np.zeros((len(eeg_channels), window_samples), dtype=np.float32)
-    filled_samples = 0
+    # Initialize empty aggregator for all EEG data
+    aggregator = np.zeros((len(eeg_channels), 0), dtype=np.float32)
 
     try:
         while True:
             try:
-                new_data = board.get_current_board_data(new_samples)
-                if new_data.size == 0 or new_data.shape[1] != new_samples:
-                    continue  # Skip if we don't have enough samples yet, e.g. at start
+                # Get a large chunk of data - board will return what it has
+                new_data = board.get_current_board_data(1024)
+                if new_data.size == 0:
+                    continue
 
-                # Extract just the EEG channels from the new data
+                # Extract EEG channels and add to aggregator
                 eeg_data = new_data[eeg_channels]
+                aggregator = np.concatenate([aggregator, eeg_data], axis=1)
 
-                ring_buffer[:, :-new_samples] = ring_buffer[:, new_samples:]
-                ring_buffer[:, -new_samples:] = eeg_data
+                # Generate windows while we have enough data
+                while aggregator.shape[1] >= window_samples:
```

Comment:
Can you explain why there is a second while loop? I naively expected the outer loop to be enough, so I must be missing something

Reply:
Old version, this no longer exists.